File: arrays/two_sum/two_sum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <limits>
using namespace std;

class Solution {
public: 
    vector<int> twoSum(vector<int>& nums, int target)
    {
        unordered_map<int,int> valueIndex;

        for(int i = 0; i < nums.size(); i++)
        {
            int diff = target - nums[i];
            if(valueIndex.count(diff) && valueIndex[diff] != i)
                return {min(i, valueIndex[diff]), max(i, valueIndex[diff])};
            else
                valueIndex[nums[i]] = i;
        }

    }
};

int main()
{
    cout << "Starting Two_Sum Solution" << endl;
    int target;
    int input;

    vector<int> n = {};
    cout << "Enter numbers (ctrl+D to quit):" << endl;
    while (std::cin >> input && input != -1) {
        n.push_back(input);
    }
    cin.clear(); 
    cin.ignore(numeric_limits<streamsize>::max(), '\n');

    cout << "You entered " << n.size() << " numbers." << endl;
    cout << "Type integer target: ";
    cin >> target;
    cout << "Integer Target is: "  << target << endl;

    Solution obj;
    vector<int> result = obj.twoSum(n, target);
    
    if(result.size() == 2)
    {
        cout << "Result: Solution Exists as follows \n";
        cout << "Indices: [" << result[0] << ", " << result[1] << "] \n";
        cout << "Values: [" << n[result[0]] << ", " << n[result[1]] << "] \n";
    }
    else    
        cout << "Result: No solution exists. \n";

    return 0;

}

File: arrays/two_sum/two_sum.py
'''
PSEUDO CODE - TWO SUM

A. HASHMAP
1. FOR array
2. if diff in map, return indices
3. else, add value:index to map

B. brute force 
1. FOR i in nums
2. FOR j in i+1 to nums
3. if nums[i] + nums[j] == target, return indices


'''
from typing import List

class Solution:

    def two_sum(self, nums: List[int], target: int) -> List[int]:

        
        result = []

        # solution 1 - brute force
        # for i in range(0, len(nums)):
        #      for j in range(i+1, len(nums)):
        #           if nums[i] + nums[j] == target:
        #                result = [ min(i,j) , max(i,j) ]
        
        # solution 2 - hashmap
        numIndex = {}
        for i in range(0, len(nums)):
            if target - nums[i] in numIndex:
                  minIndex = min(numIndex[target-nums[i]], i)
                  maxIndex = max(numIndex[target-nums[i]], i)
                  result = [minIndex, maxIndex]
            else:
                numIndex[nums[i]] = i

        return result
    

def main():
        input_array = [1,4,6,7]
        target = 10
        obj = Solution()
        result = obj.two_sum(input_array, target)
        print(result)

if __name__ == "__main__":
    main()
    
